### ISSUE TITLE
Front-end prototype / view

### ISSUE THREAD

**Issue by jamesMcMeex** [edgarfinn/Tune-matching-logic#4]
Some work is required to scope out requirements for the development of the basic view layer. Initially, I think we should focus on a few things for this:

- prototype / wireframes for the UI
- some other design stuff (starting to think we might need to do some actual “web design”?!)

We’ve got another ticket already that covers an initial high level user journey, which I think will help in achieving this prototype. After we have more clarity on the UX and design behind the prototype. I think then we need to move onto code:

- initial set up of React in the project...
- how do we handle linting of front-end code
- decide how we are going to handle CSS
- do some research on possibly other packages (for React perhaps?) that will help with interaction / animation / handling dynamic functionality

Just brain-dumping in this ticket. Fully expect this ticket to be refined / broken down / destroyed during discussion.

**Comment by edgarfinn** [comment]
I broadly agree with all of the above. 

- Can you elaborate on 'other design stuff'? Do you mean colour schemes and visual themes kind of thing? Or higher level user requirements?

Once we've prototyped the initial idea, we will hopefully have a slightly more in-depth idea of technologies, which will help inform things like project setup, styling technologies etc... 



I'm sure there will be initial project setup teething problems, Ive not started a single job or project where there aren't at least one or two. I'm sure it'll be fine though 😄 

As new requirements crop up we will need probably need to import more packages / libraries. We just need to make sure there is consensus on what we use (for example styled-components vs sass vs material-ui) **(please, not material-ui!)** and a consideration for performance (ie not hindering performance with massive libraries for small tasks). One of the most important things to me about our choice of technologies is that its driven by things we want to learn about, since that's kind of why I wanted to build this thing. So if there are styling frameworks you want to sink your teethe into with React for example I'm happy to use them. Definitely recommend `styled-components` if you want to get a taste of CSS-In-JS.

To my mind linting will be a case of us agreeing on a style-guide, one of us setting it up, and then sharing / uploading the config file.

**Comment by edgarfinn** [comment]
@jamesMcMeex Aside from building a prototype, are there any other action points you think worth putting down under this issue?

**Comment by jamesMcMeex** [comment]
> Can you elaborate on 'other design stuff'? Do you mean colour schemes and visual themes kind of thing? Or higher level user requirements?

Yes - apologies that wasn't a good description. I do mean colour schemes, fonts, border styles, form input styles, hero areas, tiles, modals / flashes, list styles, layout logic, and so on. 

### Caveat:
I'm conscious that this might be a little bit off-piste for what you might have expected to go on the project board. I am totally fine if you don't find this is very important, and you'd rather focus on the things that you really want to dive into. I'd be happy to look at this in parallel with our more functional work, and you can dip in as much as you like.

### Another caveat:
I've never done any designing, and though I don't really want to learn how to be a web designer, I think this project would benefit with just a little bit of treatment in the way of design, as it does affect the usability and the journey of the app. So, enough with the caveats...

### Component Library
I think it might be interesting to build a sort of styleguide for the app's theme. This could be a separate app in itself that presents each "component" of the UI, along with an associated design spec. It could even be extended to work with React, so we have an API ref for each React component? 

### Design System
We could look into a style system to guide us in building the UI, such as Atomic Design: http://atomicdesign.bradfrost.com/table-of-contents/. The guy who wrote Atomic Design also wrote a styleguide app called Pattern Lab: https://patternlab.io/. There are lots of similar tools and technologies that we could use though. 

### UX / User journey
I was also thinking of the user journey. I think this is quite an interesting one as I'm sure it will inform some of the functionality of the app, as the user moves through their own logic of what they want to do with the app. It will affect decisions like:

- What changes the UI
- When does the UI change
- When does data update
- How to go back to the start / reset
- What is the opening screen
- How does the user move between features
- How does the user jump to another feature

A tool like https://www.gomockingbird.com/home might be helpful here, though it's quite chunky. But as discussed, we could - maybe we should use https://pencil.evolus.vn/

Again, I've got no experience doing this, but I think maybe just a little time invested might help us iron out a process or sequence of actions that might be more confusing to puzzle out in code once we're deep into development... I dunno!

**Comment by edgarfinn** [comment]
Ive done a little bit of UX design a while back and am definitely up for trying to practise it where possible in this project. As you say, fleshing out user journeys and user stories first and foremost is a good shout. If everything works back from user journeys its much easier to keep the focus on the user's needs and priorities so Im keen to make this the central drive of the design process.

Re: web design (themes, colours etc...) as long as we dont get derailed or too side-tracked by the process then I think it could serve as a powerful tool. If its something you're passionate to explore then we should absolutely do it. My only point would be that it would be good for this to be informed where appropriate by the user stories + journeys, rather than become the driving force in the UX design that informs all else (...if that makes sense?).

FWIW I liked your vintage roland theme idea, but happy to follow your lead. There are some good-looking articles out there on colour palette design:
https://uxdesign.cc/how-to-use-color-in-ui-design-wisely-to-create-a-perfect-ui-interface-2af42f901f4

Re: component library
Im up for this in theory, but think it might be better as something we introduce later on. Reason being I think it should serve our needs (once we identify what they are) rather than determine our work flow. 

For example, once we've learnt what kind of components we use the most (ie buttons / text elements / sliders / what-have-you), we could then build the best possible component library for *our* needs. 

My concern with implementing this from the start is that it may become the tail that wags the dog.

> This could be a separate app in itself that presents each "component" of the UI, along with an associated design spec
Check out [storybook](https://storybook.js.org/). Its awesome! And if you look at most published react components (eg [airbnb's date-picker](https://github.com/airbnb/react-dates). Most of them are documented using storybook.
